File: XDrawer/Line.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing.Drawing2D;

namespace XDrawer
{
    public class Line : TwoPointFigure
    {
        public Line(Popup popup, int x, int y)
            : base(popup, x, y)
        {
        }
        public override void draw(Graphics g, Pen pen)
        {
            Color oldColor = pen.Color;
            pen.Color = _color;
            g.DrawLine(pen, _x1, _y1, _x2, _y2);
            pen.Color = oldColor;
        }
        public override void makeRegion()
        {
            int regionWidth = 20;
            int x = _x1;
            int y = _y1;
            int w = _x2 - _x1;
            int h = _y2 - _y1;
            int sign_h = 1;
            if (h < 0) sign_h = -1;
            double angle;
            double theta = (w != 0) ? Math.Atan((double)(h) / (double)(w)) : sign_h * Math.PI / 2.0;
            if (theta < 0) theta = theta + 2 * Math.PI;
            angle = (theta + Math.PI / 2.0);
            int dx = (int)(regionWidth * Math.Cos(angle));
            int dy = (int)(regionWidth * Math.Sin(angle));
            System.Drawing.Point[] pt = new System.Drawing.Point[4];
            pt[0].X = x + dx; pt[0].Y = y + dy;
            pt[1].X = x - dx; pt[1].Y = y - dy;
            pt[2].X = x + w - dx; pt[2].Y = y + h - dy;
            pt[3].X = x + w + dx; pt[3].Y = y + h + dy;
            byte[] type = new byte[4];
            type[0] = (byte)PathPointType.Line;
            type[1] = (byte)PathPointType.Line;
            type[2] = (byte)PathPointType.Line;
            type[3] = (byte)PathPointType.Line;
            GraphicsPath gp = new GraphicsPath(pt, type);
            _region = new Region(gp);
        }
    }
}

File: XDrawer/Line.cpp
#include "stdafx.h"
#include <math.h>

#include "Line.h"


#ifndef M_PI
#define M_PI (3.1415926535897)
#endif

#define sign(x) ((x>0) ? x : -x)

IMPLEMENT_SERIAL(Line, TwoPointFigure, 1)
Line::Line()
	:TwoPointFigure()
{	
}
Line::Line(int x1, int y1)
	:TwoPointFigure(x1, y1)
{
}
Line::Line(int x1, int y1, int x2, int y2)
	:TwoPointFigure(x1, y1, x2, y2)
{
}
Line::~Line()
{
}

void Line::Serialize(CArchive& ar)
{
	TwoPointFigure::Serialize(ar);
}
//Scope resolution operator
void Line::draw(CDC* pDC)
{
	CPen pen(PS_SOLID,1,_color);
	CPen *oldPen = pDC->SelectObject(&pen);
	pDC->MoveTo(m_x1, m_y1);
	pDC->LineTo(m_x2, m_y2);
	pDC->SelectObject(oldPen);
}
void Line::calcPoint()
{
	regionWidth = 15;
	x = m_x1;
	y = m_y1;
	w = m_x2 - m_x1;
	h = m_y2 - m_y1;
	theta = w ? atan(double(h)/double(w)) : sign(h)*M_PI/2.;
	if (theta < 0) theta = theta + 2 * M_PI;
	angle = (theta + M_PI / 2.);
	dx = int(regionWidth * cos(angle));
	dy = int(regionWidth * sin(angle));	
	points[0].x = x + dx;
	points[0].y = y + dy;
	points[1].x = x - dx;
	points[1].y = y - dy;
	points[2].x = x + w - dx;
	points[2].y = y + h - dy;
	points[3].x = x + w + dx;
	points[3].y = y + h + dy;
}
void Line::makeRegion()
{	
	calcPoint();
	if (region) delete region;
	region = new CRgn();
	region->CreatePolygonRgn(points,4,WINDING);
}
void Line::drawDots(CDC* pDC)
{
	calcPoint();
	dotedFlag = TRUE;	
	CRect rect1(points[0].x-DOTSIZE,points[0].y-DOTSIZE,points[0].x,points[0].y);
	CRect rect2(points[1].x, points[1].y, points[1].x+DOTSIZE, points[1].y+DOTSIZE);
	CRect rect3(points[2].x+DOTSIZE,points[2].y+DOTSIZE,points[2].x,points[2].y);
	CRect rect4(points[3].x, points[3].y, points[3].x-DOTSIZE, points[3].y-DOTSIZE);
	CBrush brush(BLACK_COLOR);
	pDC->FillRect(&rect1,&brush);
	pDC->FillRect(&rect2,&brush);
	pDC->FillRect(&rect3,&brush);
	pDC->FillRect(&rect4,&brush);
}
